docs/architecture/deep-dives/state/index.md

```diff
@@ -142,212 +143,192 @@ own key definition.
 
-## Migrating
+:::
 
-Migrating data to state providers is incredibly similar to migrating data in general. You create
-your own class that extends `Migrator<From, To>`. That will require you to implement your own
-`migrate(migrationHelper: MigrationHelper)` method. `MigrationHelper` already includes methods like
-`get` and `set` for getting and settings value to storage by their string key. There are also
-methods for getting and setting using your `KeyDefinition` or `KeyDefinitionLike` object to and from
-user and global state. An example of how you might use these new helpers is below:
+### `ActiveUserState<T>`
 
-```typescript
-type ExpectedGlobalState = { myGlobalData: string };
+:::warning
 
-type ExpectedAccountState = { myUserData: string };
+`ActiveUserState` has race condition problems. Do not use it for updates and consider transitioning
+your code to SingleUserState instead. [Read more.](#should-i-use-activeuserstate)
 
-const MY_GLOBAL_KEY_DEFINITION: KeyDefinitionLike = {
-  stateDefinition: { name: "myState" },
-  key: "myGlobalKey",
-};
-const MY_USER_KEY_DEFINITION: KeyDefinitionLike = {
-  stateDefinition: { name: "myState" },
-  key: "myUserKey",
-};
+:::
 
-export class MoveToStateProvider extends Migrator<10, 11> {
-  async migrate(migrationHelper: MigrationHelper): Promise<void> {
-    const existingGlobalData = await migrationHelper.get<ExpectedGlobalState>("global");
+`ActiveUserState<T>` is an object to help you maintain and view the state of the currently active
+user. If the currently-active user changes, like through account switching, the data this object
+represents will change along with it.
 
-    await migrationHelper.setGlobal(MY_GLOBAL_KEY_DEFINITION, {
-      myGlobalData: existingGlobalData.myGlobalData,
-    });
+### Updating state with `update`
 
-    const updateAccount = async (userId: string, account: ExpectedAccountState) => {
-      await migrationHelper.setUser(MY_USER_KEY_DEFINITION, {
-        myUserData: account.myUserData,
-      });
-    };
+The update method has options defined as follows:
 
-    const accounts = await migrationHelper.getAccounts<ExpectedAccountState>();
+```typescript
+{ActiveUser|SingleUser|Global}State<T> {
+  // ... rest of type left out for brevity
+  update<TCombine>(updateState: (state: T, dependency: TCombine) => T, options?: StateUpdateOptions);
+}
 
-    await Promise.all(accounts.map(({ userId, account }) => updateAccount(userId, account)));
-  }
+type StateUpdateOptions = {
+  shouldUpdate?: (state: T, dependency: TCombine) => boolean;
+  combineLatestWith?: Observable<TCombine>;
+  msTimeout?: number
 }
 ```
 
-:::note
-
-`getAccounts` only gets data from the legacy account object that was used in `StateService`. As data
-gets migrated off of that account object the response from `getAccounts`, which returns a record
-where the key will be a user's ID and the value being the legacy account object.
-
-:::
+:::warning `firstValueFrom()` and state updates
```

Comment:
Somewhere in here we can also mention that `update` will return the new value that was just saved back to the caller, or in the case of a custom `shouldUpdate` function that returned false it will return the existing state value. So if consumers need the updated value they can always get it that way.

Reply:
Addressed with [`fbd3861` (#549)](https://github.com/bitwarden/contributing-docs/pull/549/commits/fbd3861330cdced948f619c081bea4644cc6ed43).  Let me know if you think the warning block is too big and should just be a new section.